### PR TITLE
[PETSc] v3.21.5

### PR DESCRIPTION
**Pull request by boriskaus**
PETSc version 3.21.5, our most complete compilation so far with:
- `SuperLU_Dist`
- `MUMPS`
- `SuiteSparse`
- `Triangle`
- `TetGen`
- `HYPRE`
Note that `MPI` on windows remains broken and is deactivated.

### CODE REVIEW

File: P/PETSc/build_tarballs.jl [hunk 1]
```diff
@@ -1,23 +1,26 @@
...
 using BinaryBuilder, Pkg
 using Base.BinaryPlatforms
 const YGGDRASIL_DIR = "../.."
 include(joinpath(YGGDRASIL_DIR, "platforms", "mpi.jl"))
 
 name = "PETSc"
-version = v"3.20.5"
-petsc_version = v"3.20.5"
+version = v"3.21.4"
+petsc_version = v"3.21.4"
```
Comment (eschnett) [inline P/PETSc/build_tarballs.jl:9]: `3.21.5` is now current...

**Comment by eschnett** [comment]
I suggest setting this while building (see `Trilinos`):
```
# TODO: MPITrampoline embeds the wrong CC. https://github.com/JuliaPackaging/Yggdrasil/issues/7420
export MPITRAMPOLINE_CC="$(which $CC)"
export MPITRAMPOLINE_CXX="$(which $CXX)"
export MPITRAMPOLINE_FC="$(which $FC)"
```

Also, you can now set
```
MPItrampoline_compat_version="5.5.0"
```
and remove the line
```
platforms = filter(p -> !(p["mpi"] == "mpitrampoline" && Sys.isfreebsd(p)), platforms)
```
if you want to support MPItrampoline on BSD systems. (That's probably not important at all.)

**Comment by ViralBShah** [comment]
@eschnett It seems that after setting the things you suggested, the build is not progressing at all.

**Comment by eschnett** [comment]
Okay, I forgot to mention: `aarch64-unknown-freebsd` isn't supported yet by MPICH, you will have to exclude this explicitly until MPICH has been rebuilt https://github.com/JuliaPackaging/Yggdrasil/pull/9516 . That's unrelated to MPItrampoline. (This architecture isn't supported by OpenMPI either but should already be excluded.)

**Comment by boriskaus** [comment]
so it seems that all builds with MPItrampoline fail now...

**Comment by boriskaus** [comment]
builds currently fail because there appears to be a disk-space issue:
```
[20:42:46] Error printing error message from exception or printing the traceback:[Errno 28] No space left on device
[20:42:46] Installing license file "/workspace/srcdir/petsc-3.21.5/LICENSE" to "/workspace/destdir/share/licenses/PETSc/LICENSE"...
[20:42:46] /etc/profile.d/0_bb_utils.sh: line 35: set: write error: No space left on device
[20:42:46] /etc/profile.d/0_bb_utils.sh: line 41: echo: write error: No space left on device
[20:42:46] Child Process exited, exit code 1
ERROR: LoadError: SystemError: close: No space left on device
```

**Comment by ViralBShah** [comment]
@giordano @fredrikekre @Keno Can one of you help with the diskspace issue - or point to someone who might?

**Comment by giordano** [comment]
That was already resolved on its own (we automatically clean up agents directories at the end of the job)

**Comment by ViralBShah** [comment]
```
2024-09-26 18:10:34 EDT | [22:10:34]   443 \| PetscMPIInt PETSC_MPI_ERROR_CLASS = MPI_ERR_LASTCODE, PETSC_MPI_ERROR_CODE;
-- | --
  | 2024-09-26 18:10:34 EDT | [22:10:34]       \|                                     ^~~~~~~~~~~~~~~~
  | 2024-09-26 18:10:34 EDT | [22:10:34] 1 error generated.

```

File: P/PETSc/build_tarballs.jl [hunk 2]
```diff
@@ -280,7 +348,7 @@ build_petsc()
     if  [ "${1}" == "double" ] &&  [ "${2}" == "real" ] &&  [ "${3}" == "Int64" ] &&  [ "${4}" == "deb" ]; then
         
         # this is the example that PETSc uses to test the correct installation        
-        # We compile it with debug flags (helpful to catch issues)
+        # We compile it witn debug flags (helpful to catch issues)
```
Comment (eschnett) [inline P/PETSc/build_tarballs.jl:351]: Oops?

File: P/PETSc/build_tarballs.jl [hunk 3]
```diff
@@ -341,10 +408,13 @@ platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "armv6l" && libc(p
 platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "armv7l" && libc(p) == "glibc"), platforms)
 platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "x86_64" && libc(p) == "musl"), platforms)
 platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "i686"), platforms)
+platforms = filter(p -> !(p["mpi"] == "openmpi" && Sys.isfreebsd(p)),  platforms)
```
Comment (eschnett) [inline P/PETSc/build_tarballs.jl:411]: I suggest to add a comment here (and for MPICH below) that this excludes only `aarch64-unknown-freebsd`, and that these exclusions can go away once OpenMPI and MPICH have been built for these platforms.

File: P/PETSc/build_tarballs.jl [hunk 4]
```diff
@@ -235,6 +301,8 @@ build_petsc()
 
     # Remove PETSc.pc because petsc.pc also exists, causing conflicts on case-insensitive file-systems.
     rm ${libdir}/petsc/${PETSC_CONFIG}/lib/pkgconfig/PETSc.pc
+    # sed -i -e "s/-lpetsc/-lpetsc_${PETSC_CONFIG}/g" "$libdir/petsc/${PETSC_CONFIG}/lib/pkgconfig/petsc.pc"
```
Comment (eschnett) [inline P/PETSc/build_tarballs.jl:304]: If these lines didn't work, remove them?

File: P/PETSc/build_tarballs.jl [hunk 5]
```diff
@@ -1,23 +1,26 @@
...
 name = "PETSc"
-version = v"3.20.5"
-petsc_version = v"3.20.5"
+version = v"3.21.5"
+petsc_version = v"3.21.5"
 
-MPItrampoline_compat_version="5.2.1"
+MPItrampoline_compat_version="~5.5.0"
 MicrosoftMPI_compat_version="~10.1.4" 
 MPICH_compat_version="~4.1.2"
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:13]: Why the tilde compat?  I don't think these packages are known to break API/ABI in minor versions, are they?

File: P/PETSc/build_tarballs.jl [hunk 6]
```diff
@@ -80,8 +88,8 @@ else
 fi
 
 atomic_patch -p1 $WORKSPACE/srcdir/patches/mingw-version.patch
-atomic_patch -p1 $WORKSPACE/srcdir/patches/mpi-constants.patch
-atomic_patch -p1 $WORKSPACE/srcdir/patches/sosuffix.patch   
+atomic_patch -p1 $WORKSPACE/srcdir/patches/mpi-constants.patch
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:91]:
```suggestion
atomic_patch -p1 $WORKSPACE/srcdir/patches/mpi-constants.patch
```

File: P/PETSc/build_tarballs.jl [hunk 7]
```diff
@@ -235,7 +301,7 @@ build_petsc()
 
     # Remove PETSc.pc because petsc.pc also exists, causing conflicts on case-insensitive file-systems.
     rm ${libdir}/petsc/${PETSC_CONFIG}/lib/pkgconfig/PETSc.pc
-
+
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:304]:
```suggestion

```

File: P/PETSc/build_tarballs.jl [hunk 8]
```diff
@@ -342,9 +407,14 @@ platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "armv7l" && libc(p
 platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "x86_64" && libc(p) == "musl"), platforms)
 platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "i686"), platforms)
 
+# this excludes only aarch64-unknown-freebsd;  can be removed once OpenMPI has been built for this platforms.
+platforms = filter(p -> !(p["mpi"] == "openmpi" && Sys.isfreebsd(p)),  platforms)
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:411]: You're also excluding x86_64

File: P/PETSc/build_tarballs.jl [hunk 9]
```diff
@@ -342,9 +407,14 @@ platforms = filter(p -> !(p["mpi"] == "openmpi" && arch(p) == "armv7l" && libc(p
...
 
+# this excludes only aarch64-unknown-freebsd;  can be removed once OpenMPI has been built for this platforms.
+platforms = filter(p -> !(p["mpi"] == "openmpi" && Sys.isfreebsd(p)),  platforms)   
+
 # MPItrampoline
 platforms = filter(p -> !(p["mpi"] == "mpitrampoline" && libc(p) == "musl"), platforms)
-platforms = filter(p -> !(p["mpi"] == "mpitrampoline" && Sys.isfreebsd(p)), platforms)
+
+# MPICH
+platforms = filter(p -> !(p["mpi"] == "mpich" && Sys.isfreebsd(p)), platforms)  # can be removed once MPICH has been built for aarch64-unknown-freebsd
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:417]: Same as above.

File: P/PETSc/build_tarballs.jl (hunk 5 above)
Comment (boriskaus) [reply P/PETSc/build_tarballs.jl:13]: perhaps it is different for `MPItrampoline`, but I certainly had runtime errors with `MPICH` when the minor version was changed.

File: P/PETSc/build_tarballs.jl (hunk 6 above)
Comment (boriskaus) [reply P/PETSc/build_tarballs.jl:91]: we do need the `sosuffix` patch. I'vd removed the trailing space after the `mpi-constants.patch`

File: P/PETSc/build_tarballs.jl (hunk 8 above)
Comment (boriskaus) [reply P/PETSc/build_tarballs.jl]: changed that now

File: P/PETSc/build_tarballs.jl (hunk 9 above)
Comment (boriskaus) [reply P/PETSc/build_tarballs.jl]: same

**Comment by ViralBShah** [comment]
Should we close this in favour of the build for 3.22 in #9565

**Comment by boriskaus** [comment]
No, ideally merge both because some codes rely on a specific version of petsc

**Comment by ViralBShah** [comment]
Is it good to merge now?

**Comment by boriskaus** [comment]
Yes


Sent from my Apple Watch



On 3. Oct 2024, at 07:39, Viral B. Shah ***@***.***> wrote:


Is this one good to merge now?

**Pull request merged**